{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       "    @font-face {\n",
       "        font-family: \"Computer Modern\";\n",
       "        src: url('http://9dbb143991406a7c655e-aa5fcb0a5a4ec34cff238a2d56ca4144.r56.cf5.rackcdn.com/cmunss.otf');\n",
       "    }\n",
       "    @font-face {\n",
       "        font-family: \"Computer Modern\";\n",
       "        font-weight: bold;\n",
       "        src: url('http://9dbb143991406a7c655e-aa5fcb0a5a4ec34cff238a2d56ca4144.r56.cf5.rackcdn.com/cmunsx.otf');\n",
       "    }\n",
       "    @font-face {\n",
       "        font-family: \"Computer Modern\";\n",
       "        font-style: oblique;\n",
       "        src: url('http://9dbb143991406a7c655e-aa5fcb0a5a4ec34cff238a2d56ca4144.r56.cf5.rackcdn.com/cmunsi.otf');\n",
       "    }\n",
       "    @font-face {\n",
       "        font-family: \"Computer Modern\";\n",
       "        font-weight: bold;\n",
       "        font-style: oblique;\n",
       "        src: url('http://9dbb143991406a7c655e-aa5fcb0a5a4ec34cff238a2d56ca4144.r56.cf5.rackcdn.com/cmunso.otf');\n",
       "    }\n",
       "    h1 {\n",
       "        font-family: Helvetica, serif;\n",
       "    }\n",
       "    h4{\n",
       "        margin-top:12px;\n",
       "        margin-bottom: 3px;\n",
       "       }\n",
       "    div.text_cell_render{\n",
       "        font-family: Computer Modern, \"Helvetica Neue\", Arial, Helvetica, Geneva, sans-serif;\n",
       "        line-height: 145%;\n",
       "        font-size: 130%;\n",
       "        margin-left:0;\n",
       "        margin-right:0;\n",
       "    }\n",
       "    .CodeMirror{\n",
       "            font-family: \"Source Code Pro\", source-code-pro,Consolas, monospace;\n",
       "    }\n",
       "    .text_cell_render h5 {\n",
       "        font-weight: 300;\n",
       "        font-size: 22pt;\n",
       "        color: #4057A1;\n",
       "        font-style: italic;\n",
       "        margin-bottom: .5em;\n",
       "        margin-top: 0.5em;\n",
       "        display: block;\n",
       "    }\n",
       "    \n",
       "    .warning{\n",
       "        color: rgb( 240, 20, 20 )\n",
       "        }  \n",
       "</style>\n",
       "<script>\n",
       "    MathJax.Hub.Config({\n",
       "                        TeX: {\n",
       "                           extensions: [\"AMSmath.js\"]\n",
       "                           },\n",
       "                tex2jax: {\n",
       "                    inlineMath: [ ['$','$'], [\"\\\\(\",\"\\\\)\"] ],\n",
       "                    displayMath: [ ['$$','$$'], [\"\\\\[\",\"\\\\]\"] ]\n",
       "                },\n",
       "                displayAlign: 'center', // Change this to 'center' to center equations.\n",
       "                \"HTML-CSS\": {\n",
       "                    styles: {'.MathJax_Display': {\"margin\": 4}}\n",
       "                }\n",
       "        });\n",
       "</script>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.core.display import HTML\n",
    "from IPython.lib.display import YouTubeVideo\n",
    "from IPython.display import Image\n",
    "\n",
    "def css_styling():\n",
    "    styles = open(\"custom.css\", \"r\").read()\n",
    "    return HTML(styles)\n",
    "css_styling()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Intro to gene set enrichment analysis (GSEA)\n",
    "### Written by Reese Richardson for use in Biol Sci 378, Winter 2022, Northwestern University (rakr@u.northwestern.edu)\n",
    "\n",
    "Welcome back to Python! First, let's import the libraries that we'll need."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import scipy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4.1 The fundamentals of GSEA\n",
    "\n",
    "Consider the humble [tribble](https://en.wikipedia.org/wiki/Tribble), an organism with only 300 genes. \n",
    "\n",
    "Say we perform a gene expression experiment in the tribble's thymus before and after repeated tachyon pulse exposure. We have 10 replicates each in the control ('control') and treatment ('tachyon') group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "tribble_df = pd.read_csv('tribble_expression_experiment_220217.csv')\n",
    "tribble_df = tribble_df.set_index('gene_name')\n",
    "sample_names = tribble_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>control_0</th>\n",
       "      <th>control_1</th>\n",
       "      <th>control_2</th>\n",
       "      <th>control_3</th>\n",
       "      <th>control_4</th>\n",
       "      <th>control_5</th>\n",
       "      <th>control_6</th>\n",
       "      <th>control_7</th>\n",
       "      <th>control_8</th>\n",
       "      <th>control_9</th>\n",
       "      <th>tachyon_0</th>\n",
       "      <th>tachyon_1</th>\n",
       "      <th>tachyon_2</th>\n",
       "      <th>tachyon_3</th>\n",
       "      <th>tachyon_4</th>\n",
       "      <th>tachyon_5</th>\n",
       "      <th>tachyon_6</th>\n",
       "      <th>tachyon_7</th>\n",
       "      <th>tachyon_8</th>\n",
       "      <th>tachyon_9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>XPQ5</th>\n",
       "      <td>664.698874</td>\n",
       "      <td>408.999121</td>\n",
       "      <td>456.243605</td>\n",
       "      <td>667.688929</td>\n",
       "      <td>377.236813</td>\n",
       "      <td>1014.659688</td>\n",
       "      <td>1685.969741</td>\n",
       "      <td>651.426939</td>\n",
       "      <td>1602.633604</td>\n",
       "      <td>1290.682237</td>\n",
       "      <td>112.592865</td>\n",
       "      <td>76.267275</td>\n",
       "      <td>232.462429</td>\n",
       "      <td>161.237868</td>\n",
       "      <td>103.494172</td>\n",
       "      <td>38.565015</td>\n",
       "      <td>71.084933</td>\n",
       "      <td>87.813068</td>\n",
       "      <td>147.149214</td>\n",
       "      <td>46.207763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCO3</th>\n",
       "      <td>15.088205</td>\n",
       "      <td>23.592555</td>\n",
       "      <td>18.903970</td>\n",
       "      <td>15.179185</td>\n",
       "      <td>14.035329</td>\n",
       "      <td>55.304984</td>\n",
       "      <td>11.271889</td>\n",
       "      <td>16.034206</td>\n",
       "      <td>12.305053</td>\n",
       "      <td>11.262950</td>\n",
       "      <td>56.150492</td>\n",
       "      <td>11.909692</td>\n",
       "      <td>34.762178</td>\n",
       "      <td>26.202282</td>\n",
       "      <td>62.724094</td>\n",
       "      <td>22.364778</td>\n",
       "      <td>21.742087</td>\n",
       "      <td>26.614903</td>\n",
       "      <td>30.448945</td>\n",
       "      <td>42.879803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>QJW3</th>\n",
       "      <td>18.908650</td>\n",
       "      <td>13.975928</td>\n",
       "      <td>12.075902</td>\n",
       "      <td>36.342257</td>\n",
       "      <td>12.838709</td>\n",
       "      <td>17.388032</td>\n",
       "      <td>39.347269</td>\n",
       "      <td>28.799523</td>\n",
       "      <td>23.125428</td>\n",
       "      <td>19.048399</td>\n",
       "      <td>58.576843</td>\n",
       "      <td>41.233798</td>\n",
       "      <td>41.946285</td>\n",
       "      <td>48.769722</td>\n",
       "      <td>83.059179</td>\n",
       "      <td>49.303907</td>\n",
       "      <td>45.016567</td>\n",
       "      <td>66.207978</td>\n",
       "      <td>53.916145</td>\n",
       "      <td>45.274500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EBN6</th>\n",
       "      <td>20.402094</td>\n",
       "      <td>44.604160</td>\n",
       "      <td>15.125772</td>\n",
       "      <td>41.111830</td>\n",
       "      <td>23.420009</td>\n",
       "      <td>35.232104</td>\n",
       "      <td>86.254813</td>\n",
       "      <td>137.936276</td>\n",
       "      <td>36.559704</td>\n",
       "      <td>27.163797</td>\n",
       "      <td>54.705556</td>\n",
       "      <td>35.897614</td>\n",
       "      <td>59.157124</td>\n",
       "      <td>55.043423</td>\n",
       "      <td>80.384513</td>\n",
       "      <td>42.614611</td>\n",
       "      <td>29.159306</td>\n",
       "      <td>38.294632</td>\n",
       "      <td>73.466202</td>\n",
       "      <td>31.787099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HRP2</th>\n",
       "      <td>11.797595</td>\n",
       "      <td>14.747975</td>\n",
       "      <td>6.922999</td>\n",
       "      <td>11.931877</td>\n",
       "      <td>20.730447</td>\n",
       "      <td>15.115598</td>\n",
       "      <td>12.558447</td>\n",
       "      <td>25.923035</td>\n",
       "      <td>22.902441</td>\n",
       "      <td>7.050126</td>\n",
       "      <td>221.323388</td>\n",
       "      <td>170.403353</td>\n",
       "      <td>69.209947</td>\n",
       "      <td>120.267446</td>\n",
       "      <td>119.176199</td>\n",
       "      <td>87.188052</td>\n",
       "      <td>58.144189</td>\n",
       "      <td>85.320047</td>\n",
       "      <td>204.856581</td>\n",
       "      <td>105.498654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECA7</th>\n",
       "      <td>18022.540668</td>\n",
       "      <td>13547.763889</td>\n",
       "      <td>37731.397132</td>\n",
       "      <td>34926.324200</td>\n",
       "      <td>31657.321653</td>\n",
       "      <td>26417.159277</td>\n",
       "      <td>31850.863034</td>\n",
       "      <td>6461.423038</td>\n",
       "      <td>14014.375500</td>\n",
       "      <td>7321.074454</td>\n",
       "      <td>30010.884652</td>\n",
       "      <td>18018.552133</td>\n",
       "      <td>14729.356321</td>\n",
       "      <td>9202.711608</td>\n",
       "      <td>17010.133680</td>\n",
       "      <td>10562.769632</td>\n",
       "      <td>11720.014148</td>\n",
       "      <td>28667.467248</td>\n",
       "      <td>11884.578586</td>\n",
       "      <td>4073.747175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RST6</th>\n",
       "      <td>3082.033135</td>\n",
       "      <td>9016.571493</td>\n",
       "      <td>4747.837657</td>\n",
       "      <td>3436.169598</td>\n",
       "      <td>2448.532295</td>\n",
       "      <td>3680.932390</td>\n",
       "      <td>1157.172271</td>\n",
       "      <td>6137.881230</td>\n",
       "      <td>4602.268527</td>\n",
       "      <td>5868.060649</td>\n",
       "      <td>5400.560156</td>\n",
       "      <td>2873.010871</td>\n",
       "      <td>1416.955803</td>\n",
       "      <td>7034.156555</td>\n",
       "      <td>1859.256229</td>\n",
       "      <td>6878.963535</td>\n",
       "      <td>4386.597447</td>\n",
       "      <td>10267.250768</td>\n",
       "      <td>5699.904478</td>\n",
       "      <td>3761.187693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAA9</th>\n",
       "      <td>54.991758</td>\n",
       "      <td>85.084498</td>\n",
       "      <td>39.424749</td>\n",
       "      <td>68.502921</td>\n",
       "      <td>46.407044</td>\n",
       "      <td>53.719819</td>\n",
       "      <td>34.833787</td>\n",
       "      <td>53.273648</td>\n",
       "      <td>48.205888</td>\n",
       "      <td>81.070673</td>\n",
       "      <td>298.427523</td>\n",
       "      <td>703.861536</td>\n",
       "      <td>316.959462</td>\n",
       "      <td>859.621810</td>\n",
       "      <td>388.539778</td>\n",
       "      <td>414.893759</td>\n",
       "      <td>266.628187</td>\n",
       "      <td>734.879757</td>\n",
       "      <td>654.941683</td>\n",
       "      <td>1242.523123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VFS4</th>\n",
       "      <td>4.675148</td>\n",
       "      <td>10.066881</td>\n",
       "      <td>15.947455</td>\n",
       "      <td>8.702472</td>\n",
       "      <td>10.972780</td>\n",
       "      <td>9.477237</td>\n",
       "      <td>17.384920</td>\n",
       "      <td>31.543625</td>\n",
       "      <td>15.916543</td>\n",
       "      <td>21.067065</td>\n",
       "      <td>5.287752</td>\n",
       "      <td>7.043179</td>\n",
       "      <td>5.640207</td>\n",
       "      <td>3.589583</td>\n",
       "      <td>3.923198</td>\n",
       "      <td>7.019026</td>\n",
       "      <td>7.564316</td>\n",
       "      <td>8.533002</td>\n",
       "      <td>7.081291</td>\n",
       "      <td>3.989328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAJ8</th>\n",
       "      <td>48.109166</td>\n",
       "      <td>32.923594</td>\n",
       "      <td>40.740428</td>\n",
       "      <td>46.305200</td>\n",
       "      <td>27.876503</td>\n",
       "      <td>65.164967</td>\n",
       "      <td>57.849654</td>\n",
       "      <td>99.034790</td>\n",
       "      <td>40.336453</td>\n",
       "      <td>242.132136</td>\n",
       "      <td>39.591045</td>\n",
       "      <td>15.545519</td>\n",
       "      <td>30.721731</td>\n",
       "      <td>16.644615</td>\n",
       "      <td>16.726740</td>\n",
       "      <td>20.374746</td>\n",
       "      <td>33.407522</td>\n",
       "      <td>22.348478</td>\n",
       "      <td>51.411569</td>\n",
       "      <td>18.163993</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>300 rows Ã— 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              control_0     control_1     control_2     control_3  \\\n",
       "gene_name                                                           \n",
       "XPQ5         664.698874    408.999121    456.243605    667.688929   \n",
       "CCO3          15.088205     23.592555     18.903970     15.179185   \n",
       "QJW3          18.908650     13.975928     12.075902     36.342257   \n",
       "EBN6          20.402094     44.604160     15.125772     41.111830   \n",
       "HRP2          11.797595     14.747975      6.922999     11.931877   \n",
       "...                 ...           ...           ...           ...   \n",
       "ECA7       18022.540668  13547.763889  37731.397132  34926.324200   \n",
       "RST6        3082.033135   9016.571493   4747.837657   3436.169598   \n",
       "FAA9          54.991758     85.084498     39.424749     68.502921   \n",
       "VFS4           4.675148     10.066881     15.947455      8.702472   \n",
       "NAJ8          48.109166     32.923594     40.740428     46.305200   \n",
       "\n",
       "              control_4     control_5     control_6    control_7  \\\n",
       "gene_name                                                          \n",
       "XPQ5         377.236813   1014.659688   1685.969741   651.426939   \n",
       "CCO3          14.035329     55.304984     11.271889    16.034206   \n",
       "QJW3          12.838709     17.388032     39.347269    28.799523   \n",
       "EBN6          23.420009     35.232104     86.254813   137.936276   \n",
       "HRP2          20.730447     15.115598     12.558447    25.923035   \n",
       "...                 ...           ...           ...          ...   \n",
       "ECA7       31657.321653  26417.159277  31850.863034  6461.423038   \n",
       "RST6        2448.532295   3680.932390   1157.172271  6137.881230   \n",
       "FAA9          46.407044     53.719819     34.833787    53.273648   \n",
       "VFS4          10.972780      9.477237     17.384920    31.543625   \n",
       "NAJ8          27.876503     65.164967     57.849654    99.034790   \n",
       "\n",
       "              control_8    control_9     tachyon_0     tachyon_1  \\\n",
       "gene_name                                                          \n",
       "XPQ5        1602.633604  1290.682237    112.592865     76.267275   \n",
       "CCO3          12.305053    11.262950     56.150492     11.909692   \n",
       "QJW3          23.125428    19.048399     58.576843     41.233798   \n",
       "EBN6          36.559704    27.163797     54.705556     35.897614   \n",
       "HRP2          22.902441     7.050126    221.323388    170.403353   \n",
       "...                 ...          ...           ...           ...   \n",
       "ECA7       14014.375500  7321.074454  30010.884652  18018.552133   \n",
       "RST6        4602.268527  5868.060649   5400.560156   2873.010871   \n",
       "FAA9          48.205888    81.070673    298.427523    703.861536   \n",
       "VFS4          15.916543    21.067065      5.287752      7.043179   \n",
       "NAJ8          40.336453   242.132136     39.591045     15.545519   \n",
       "\n",
       "              tachyon_2    tachyon_3     tachyon_4     tachyon_5  \\\n",
       "gene_name                                                          \n",
       "XPQ5         232.462429   161.237868    103.494172     38.565015   \n",
       "CCO3          34.762178    26.202282     62.724094     22.364778   \n",
       "QJW3          41.946285    48.769722     83.059179     49.303907   \n",
       "EBN6          59.157124    55.043423     80.384513     42.614611   \n",
       "HRP2          69.209947   120.267446    119.176199     87.188052   \n",
       "...                 ...          ...           ...           ...   \n",
       "ECA7       14729.356321  9202.711608  17010.133680  10562.769632   \n",
       "RST6        1416.955803  7034.156555   1859.256229   6878.963535   \n",
       "FAA9         316.959462   859.621810    388.539778    414.893759   \n",
       "VFS4           5.640207     3.589583      3.923198      7.019026   \n",
       "NAJ8          30.721731    16.644615     16.726740     20.374746   \n",
       "\n",
       "              tachyon_6     tachyon_7     tachyon_8    tachyon_9  \n",
       "gene_name                                                         \n",
       "XPQ5          71.084933     87.813068    147.149214    46.207763  \n",
       "CCO3          21.742087     26.614903     30.448945    42.879803  \n",
       "QJW3          45.016567     66.207978     53.916145    45.274500  \n",
       "EBN6          29.159306     38.294632     73.466202    31.787099  \n",
       "HRP2          58.144189     85.320047    204.856581   105.498654  \n",
       "...                 ...           ...           ...          ...  \n",
       "ECA7       11720.014148  28667.467248  11884.578586  4073.747175  \n",
       "RST6        4386.597447  10267.250768   5699.904478  3761.187693  \n",
       "FAA9         266.628187    734.879757    654.941683  1242.523123  \n",
       "VFS4           7.564316      8.533002      7.081291     3.989328  \n",
       "NAJ8          33.407522     22.348478     51.411569    18.163993  \n",
       "\n",
       "[300 rows x 20 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tribble_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's calculate the fold changes of each of these genes in the tachyon condition vs the control condition and perform a t-test to determine which genes are differentially expressed. We won't perform multiple hypothesis correction on our t-tests, as for GSEA, we will mostly be interested in just fold change. We'll visualize the results in a volcano plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# label our data\n",
    "labels = np.array(['control']*10 + ['tachyon']*10)\n",
    "\n",
    "# take our control and tachyon data\n",
    "control_vals = tribble_df[sample_names[labels == 'control']].values\n",
    "treatment_vals = tribble_df[sample_names[labels == 'tachyon']].values\n",
    "# perform a t-test\n",
    "tribble_df['ttest_p'] = scipy.stats.ttest_ind(control_vals.T,treatment_vals.T, equal_var=False).pvalue\n",
    "# calculate log2 fold change (log2(mean(tachyon)/mean(control)))\n",
    "tribble_df['log2FC'] = np.log2(np.mean(treatment_vals, axis=1)/np.mean(control_vals, axis=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(5,5))\n",
    "plt.scatter(tribble_df['log2FC'], -np.log10(tribble_df['ttest_p']))\n",
    "\n",
    "ax = plt.gca()\n",
    "ax.set_xlim([-6,6])\n",
    "ax.set_ylim([0,7])\n",
    "ax.axvline(0, color='k', linewidth=0.5)\n",
    "ax.tick_params(labelsize=14) # set the fontsize of the tick labels\n",
    "ax.set_xlabel(r'log$_2$(FC)', fontsize=16)\n",
    "ax.set_ylabel(r'-log$_{10}$(pval)', fontsize=16)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the plot above, the genes that are measured to be higher-expressed in the tachyon condition are on the right (positive values), and the genes that are higher-expressed in the control condition are on the left (negative values).\n",
    "\n",
    "Now, we get to our question of interest: **is the DNA repair pathway up-regulated in the tribble thymus after tachyon pulse exposure?**\n",
    "\n",
    "I've provided the consistuent members of the tribble DNA repair pathway in the variable `pathway`. Let's visualize where those genes fall on the volcano plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathway = np.array(['DDV5', 'JND7', 'CVT6', 'PVZ6', 'PMN6', 'JMI1', 'XHQ3', 'HLE8',\n",
    "       'HPC1', 'OCU4', 'FAA9', 'HPC1', 'JIW3', 'AUQ3', 'HMG6', 'SCU4',\n",
    "       'OEJ5', 'TET8', 'DUK9', 'WKP7', 'TCL8', 'VXN7', 'BYT7', 'FAA9',\n",
    "       'NMZ5', 'BYT7', 'JOD7', 'RQP8', 'ZXT6', 'PQJ2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(pathway)\n",
    "fig = plt.figure(figsize=(5,5))\n",
    "plt.scatter(tribble_df['log2FC'], -np.log10(tribble_df['ttest_p']))\n",
    "tribble_df_slice = tribble_df[tribble_df.index.isin(pathway)]\n",
    "plt.scatter(tribble_df_slice['log2FC'], -np.log10(tribble_df_slice['ttest_p']), label='DNA repair pathway')\n",
    "\n",
    "ax = plt.gca()\n",
    "ax.set_xlim([-6,6])\n",
    "ax.set_ylim([0,8])\n",
    "ax.legend(fontsize=16)\n",
    "ax.axvline(0, color='k', linewidth=0.5)\n",
    "ax.tick_params(labelsize=14) # set the fontsize of the tick labels\n",
    "ax.set_xlabel(r'log$_2$(FC)', fontsize=16)\n",
    "ax.set_ylabel(r'-log$_{10}$(pval)', fontsize=16)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Well...maybe? Of the 30 genes that compose the DNA repair pathway, a majority of them appear to be up-regulated in the tachyon condition. However, we should search for a more quantitative answer than this. To do so, we will use a method known as [**Gene Set Enrichment Analysis**](https://www.pnas.org/content/102/43/15545).\n",
    "\n",
    "To start, let's order our genes from highest (most positive) fold-change to lowest (most negative) fold change."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tribble_df_sorted = tribble_df.sort_values('log2FC', ascending=False)\n",
    "tribble_df_sorted"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll also create an array of length 300 (the number of genes in our organism), that shows which genes (after the genes are sorted by fold change) appear in our pathway. We'll store this in the variable `barcode`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode = tribble_df_sorted.index.isin(pathway)\n",
    "print(barcode)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll visualized the fold changes for each gene as a line plot and the barcode as, well, a barcode. In this barcode, a black cell means that gene is present in the pathway, whereas a white cell means that gene is not present in the pathway."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=2, ncols=1, sharex=True, figsize=(8,5), gridspec_kw={'height_ratios': [3, 1]})\n",
    "\n",
    "axes[0].plot(np.arange(300)+0.5, tribble_df_sorted['log2FC'].values)\n",
    "axes[0].axhline(0, color='k', linewidth=0.5)\n",
    "axes[0].set_ylabel(r'log$_2$(FC)', fontsize=16)\n",
    "axes[1].imshow(barcode.reshape(-1,1).T, aspect='auto', cmap='binary')\n",
    "axes[1].set_yticks([])\n",
    "axes[1].set_xlabel('gene rank', fontsize=16)\n",
    "axes[1].set_xlim([0,300])\n",
    "\n",
    "for ax in axes:\n",
    "    ax.tick_params(labelsize=14) # set the fontsize of the tick labels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From the barcode, we get the same impression we got from our volcano plot: the genes in the DNA repair pathway are mostly among the up-regulated genes. But let's get more quantitative:\n",
    "\n",
    "Imagine that we are walking along the barcode from front to back and taking a running sum. We stop at each element of the barcode (i.e. each gene). Starting with a running sum of zero, when we encounter a gene that is in our pathway, we'll add $1/30$ to our running sum (1 over the number of genes in our pathway). When we encounter a gene that is *not* in our pathway, we'll subtract $1/270$ from our running sum (1 over the number of genes *not* in our pathway)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "running_sum_storage = []\n",
    "running_sum = 0\n",
    "running_sum_storage.append(running_sum)\n",
    "n_in_pathway = np.sum(barcode)\n",
    "n_not_in_pathway = len(barcode) - np.sum(barcode)\n",
    "for ele in barcode:\n",
    "    if ele: # gene is in gene set\n",
    "        running_sum += 1/n_in_pathway\n",
    "    else: # gene is not in gene set\n",
    "        running_sum -= 1/n_not_in_pathway\n",
    "    running_sum_storage.append(running_sum)\n",
    "running_sum_storage = np.array(running_sum_storage)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's now plot `running_sum_storage` versus the barcode. This running sum is known as our **running enrichment score**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=3, ncols=1, sharex=True, figsize=(5,6), gridspec_kw={'height_ratios': [3,2, 1]})\n",
    "\n",
    "axes[0].plot(running_sum_storage, color='lime')\n",
    "axes[0].axhline(0.0, color='k', linewidth=0.5)\n",
    "axes[0].set_ylim([-np.max(np.abs(running_sum_storage))*1.1, np.max(np.abs(running_sum_storage))*1.1])\n",
    "axes[0].set_ylabel('enrichment score', fontsize=16)\n",
    "axes[1].plot(np.arange(300)+0.5, tribble_df_sorted['log2FC'].values)\n",
    "axes[1].axhline(0, color='k', linewidth=0.5)\n",
    "axes[1].set_ylabel(r'log$_2$(FC)', fontsize=16)\n",
    "axes[2].imshow(tribble_df_sorted.index.isin(pathway).reshape(-1,1).T, aspect='auto', cmap='binary')\n",
    "axes[2].set_yticks([])\n",
    "axes[2].set_xlabel('gene index', fontsize=16)\n",
    "axes[2].set_xlim([0,300])\n",
    "\n",
    "for ax in axes:\n",
    "    ax.tick_params(labelsize=14) # set the fontsize of the tick labels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You'll see that our enrichment score trace starts at 0 and ends at 0, but peaks at some point in the middle. The point at which our running enrichment score reaches its highest deviation from zero is our pathway's overall **Enrichment Score (ES)**. For this pathway, we find an enrichment score of..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "es_true = running_sum_storage[np.argmax(np.abs(running_sum_storage))]\n",
    "print(es_true)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that if our pathway's genes are mostly among our comparison's up-regulated genes, our enrichment score will be positive, as in the example below. However, if our pathway's genes are mostly among our comparison's down-regulated genes (such as in the hair synthesis pathway, stored in  `new_pathway`), our enrichment score will be negative, as in the example below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_pathway = np.array(['JQO7', 'PDX4', 'QLG5', 'RDD7', 'LZK3', 'RPV6', 'TVO1', 'ZXT6',\n",
    "       'XTF6', 'NXB4', 'VFS4', 'XTF6', 'QSV2', 'JPB9', 'DBS1', 'LHE7',\n",
    "       'ZAG1', 'OGU7', 'AKV4', 'SSI9', 'BNF3', 'GEW2', 'BUS2', 'VFS4',\n",
    "       'KYJ7', 'BUS2', 'ZXT6', 'DJQ5', 'ERR1', 'CDG2'])\n",
    "\n",
    "barcode = tribble_df_sorted.index.isin(new_pathway)\n",
    "\n",
    "running_sum_storage = []\n",
    "running_sum = 0\n",
    "running_sum_storage.append(running_sum)\n",
    "n_in_pathway = np.sum(barcode)\n",
    "n_not_in_pathway = len(barcode) - np.sum(barcode)\n",
    "for ele in barcode:\n",
    "    if ele: # gene is in gene set\n",
    "        running_sum += 1/n_in_pathway\n",
    "    else: # gene is not in gene set\n",
    "        running_sum -= 1/n_not_in_pathway\n",
    "    running_sum_storage.append(running_sum)\n",
    "running_sum_storage = np.array(running_sum_storage)\n",
    "\n",
    "fig, axes = plt.subplots(nrows=3, ncols=1, sharex=True, figsize=(5,6), gridspec_kw={'height_ratios': [3,2, 1]})\n",
    "\n",
    "axes[0].plot(running_sum_storage, color='lime')\n",
    "axes[0].axhline(0.0, color='k', linewidth=0.5)\n",
    "axes[0].set_ylim([-np.max(np.abs(running_sum_storage))*1.1, np.max(np.abs(running_sum_storage))*1.1])\n",
    "axes[0].set_ylabel('enrichment score', fontsize=16)\n",
    "axes[1].plot(np.arange(300)+0.5, tribble_df_sorted['log2FC'].values)\n",
    "axes[1].axhline(0, color='k', linewidth=0.5)\n",
    "axes[1].set_ylabel(r'log$_2$(FC)', fontsize=16)\n",
    "axes[2].imshow(barcode.reshape(-1,1).T, aspect='auto', cmap='binary')\n",
    "axes[2].set_yticks([])\n",
    "axes[2].set_xlabel('gene index', fontsize=16)\n",
    "axes[2].set_xlim([0,300])\n",
    "\n",
    "for ax in axes:\n",
    "    ax.tick_params(labelsize=14) # set the fontsize of the tick labels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the DNA repair pathway, we find our enrichment score to be positive! Thus, you might think that our pathway is up-regulated (AKA \"enriched\").\n",
    "\n",
    "Let's not warp jump to conclusions! We want to know whether or not this enrichment is of any statistical significance. Thus, we should compare how this enrichment score compares to the distribution of enrichment scores we would expect by random chance alone (AKA the null the distribution).\n",
    "\n",
    "But how do we generate this null distribution of enrichment scores?\n",
    "\n",
    "### 4.1.1 GSEA with phenotype permutation\n",
    "\n",
    "Let's perform the same procedure for calculating the enrichment score, but this time, before we calculate fold changes between the two conditions, we'll randomly shuffle (or \"permute\") the labels on our data so that some of the true control groups are labeled as treatment groups and some of the treatment groups are labeled as control."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(1701)\n",
    "\n",
    "# randomly permute our labels\n",
    "new_labels = np.random.permutation(labels)\n",
    "\n",
    "control_vals = tribble_df[sample_names[new_labels == 'control']].values\n",
    "treatment_vals = tribble_df[sample_names[new_labels == 'tachyon']].values\n",
    "\n",
    "# calculate log2 fold change (log2(mean(tachyon)/mean(control)))\n",
    "tribble_df['log2FC'] = np.log2(np.mean(treatment_vals, axis=1)/np.mean(control_vals, axis=1))\n",
    "\n",
    "tribble_df_sorted = tribble_df.sort_values('log2FC', ascending=False)\n",
    "\n",
    "barcode = tribble_df_sorted.index.isin(pathway)\n",
    "\n",
    "running_sum_storage = []\n",
    "running_sum = 0\n",
    "running_sum_storage.append(running_sum)\n",
    "n_in_pathway = np.sum(barcode)\n",
    "n_not_in_pathway = len(barcode) - np.sum(barcode)\n",
    "for ele in barcode:\n",
    "    if ele: # gene is in gene set\n",
    "        running_sum += 1/n_in_pathway\n",
    "    else: # gene is not in gene set\n",
    "        running_sum -= 1/n_not_in_pathway\n",
    "    running_sum_storage.append(running_sum)\n",
    "running_sum_storage = np.array(running_sum_storage)\n",
    "\n",
    "fig, axes = plt.subplots(nrows=3, ncols=1, sharex=True, figsize=(5,6), gridspec_kw={'height_ratios': [3,2, 1]})\n",
    "\n",
    "axes[0].plot(running_sum_storage, color='lime')\n",
    "axes[0].axhline(0.0, color='k', linewidth=0.5)\n",
    "axes[0].set_ylim([-np.max(np.abs(running_sum_storage))*1.1, np.max(np.abs(running_sum_storage))*1.1])\n",
    "axes[0].set_ylabel('enrichment score', fontsize=16)\n",
    "axes[1].plot(np.arange(300)+0.5, tribble_df_sorted['log2FC'].values)\n",
    "axes[1].axhline(0, color='k', linewidth=0.5)\n",
    "axes[1].set_ylabel(r'log$_2$(FC)', fontsize=16)\n",
    "axes[2].imshow(tribble_df_sorted.index.isin(pathway).reshape(-1,1).T, aspect='auto', cmap='binary')\n",
    "axes[2].set_yticks([])\n",
    "axes[2].set_xlabel('gene index', fontsize=16)\n",
    "axes[2].set_xlim([0,300])\n",
    "\n",
    "for ax in axes:\n",
    "    ax.tick_params(labelsize=14) # set the fontsize of the tick labels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we randomly permute the labels of our data (AKA the \"phenotype\" of each sample), we get a lower enrichment score than before. If we permute 1000 times, what distribution of enrichment scores do we get? Let's find out!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(1701)\n",
    "\n",
    "n_permutations = 1000\n",
    "es_null_storage_array = []\n",
    "for permute_n in range(n_permutations):\n",
    "    new_labels = np.random.permutation(labels)\n",
    "\n",
    "    control_vals = tribble_df[sample_names[new_labels == 'control']].values\n",
    "    treatment_vals = tribble_df[sample_names[new_labels == 'tachyon']].values\n",
    "\n",
    "    # calculate log2 fold change (log2(mean(tachyon)/mean(control)))\n",
    "    tribble_df['log2FC'] = np.log2(np.mean(treatment_vals, axis=1)/np.mean(control_vals, axis=1))\n",
    "\n",
    "    tribble_df_sorted = tribble_df.sort_values('log2FC', ascending=False)\n",
    "\n",
    "    barcode = tribble_df_sorted.index.isin(pathway)\n",
    "\n",
    "    running_sum_storage = []\n",
    "    running_sum = 0\n",
    "    running_sum_storage.append(running_sum)\n",
    "    n_in_pathway = np.sum(barcode)\n",
    "    n_not_in_pathway = len(barcode) - np.sum(barcode)\n",
    "    for ele in barcode:\n",
    "        if ele: # gene is in gene set\n",
    "            running_sum += 1/n_in_pathway\n",
    "        else: # gene is not in gene set\n",
    "            running_sum -= 1/n_not_in_pathway\n",
    "        running_sum_storage.append(running_sum)\n",
    "    running_sum_storage = np.array(running_sum_storage)\n",
    "    # get value of max deviation from zero\n",
    "    es_null_storage_array.append(running_sum_storage[np.argmax(np.abs(running_sum_storage))])\n",
    "\n",
    "es_null_storage_array = np.array(es_null_storage_array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(8,5))\n",
    "sns.distplot(es_null_storage_array, bins=np.arange(-1,1.01,0.1))\n",
    "\n",
    "ax = plt.gca()\n",
    "ax.axvline(es_true, color='r', label='true ES')\n",
    "ax.legend(fontsize=14)\n",
    "ax.set_xlabel('ES (null)', fontsize=16)\n",
    "ax.set_ylabel('density', fontsize=16)\n",
    "ax.tick_params(labelsize=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When we plot the distribution of expected enrichments scores, we find a symmetric bimodal distribution, and find that our true enrichment score (`es_true`) is well higher than the vast majority of enrichment scores. Among the positive enrichment scores from the null, the probability of finding an ES greater than or equal to our true enrichment score is..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_es = es_null_storage_array[es_null_storage_array > 0] # get array of positive enrichment scores\n",
    "print(np.sum(pos_es >= es_true)/len(pos_es))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is our **empirical p-value**. And thus, we've determined significant enrichment! In other words, at p < 0.05, we find the DNA repair pathway to be significantly up-regulated.\n",
    "\n",
    "The distribution of enrichment scores from the null will change based on the size of the pathway you test and the number of genes in your gene set. We can normalize our true enrichment score by dividing it by the mean of the absolute value of null enrichment scores of the same sign (i.e. if our enrichment score is negative, divide our enrichment score by the absolute value of the mean of the negative enrichment scores from the null). This quantity is known as the **Normalized Enrichment Score (NES)**. If you are performing GSEA on multiple gene sets, the NES allows you to compare which gene sets are \"more enriched\" than the others. The [original GSEA paper](https://www.pnas.org/content/102/43/15545) recommends normalizing each null enrichment score across all gene sets and using your computed true NES relative to this distribution to calculate q-values (i.e. to estimate the FDR). This is generally how multiple-hypothesis correction is performed in GSEA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nes = es_true/np.mean(np.abs(pos_es))\n",
    "print(nes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.1.2 GSEA with gene permutation\n",
    "\n",
    "When we permute our labels with $k$ replicates in our treatment condition out of $n$ samples in total, there are \n",
    "\n",
    "$${n \\choose k}$$ \n",
    "\n",
    "different arrangements of labels we can make. For $k = 10$ and $n = 20$, as we've explored above, we find that \n",
    "\n",
    "$${n \\choose k} = 184,756.$$ \n",
    "\n",
    "Hence, we can compute 184,756 unique null enrichment scores. However, if we had far fewer replicates, we couldn't make as many unique enrichment scores and thus would wind up with some repeated null observations when we permute a finite number of times. For instance, if $k = 3$ and $n = 7$, we find that \n",
    "\n",
    "$${n \\choose k} = 35.$$ \n",
    "\n",
    "Thirty-five unique null enrichment scores is not really enough to create a suitable null distribution from which we can determine an empirical p-value!\n",
    "\n",
    "If you find yourself in this situation, you should first consider whether or not your RNA-seq experiment is adequately statistically powered or if you need to run more replicates. Once you've decided to ignore how underpowered your RNA-seq experiment is, you can perform GSEA with gene permutation.\n",
    "\n",
    "In this variation of GSEA, instead of permuting phenotype labels and recalculating fold-changes, we stick with the fold changes we obtain with the correct phenotype labels and instead permute the set of genes that are in our pathway (i.e. scramble our barcode). In other words, for a gene set of size $m$, each random permutation will select $m$ random genes. The cell below demonstrates this method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "####### Finding true ES\n",
    "\n",
    "# label our data\n",
    "labels = np.array(['control']*10 + ['tachyon']*10)\n",
    "\n",
    "# take our control and tachyon data\n",
    "control_vals = tribble_df[sample_names[labels == 'control']].values\n",
    "treatment_vals = tribble_df[sample_names[labels == 'tachyon']].values\n",
    "\n",
    "# calculate log2 fold change (log2(mean(tachyon)/mean(control)))\n",
    "tribble_df['log2FC'] = np.log2(np.mean(treatment_vals, axis=1)/np.mean(control_vals, axis=1))\n",
    "\n",
    "tribble_df_sorted = tribble_df.sort_values('log2FC', ascending=False)\n",
    "\n",
    "barcode = tribble_df_sorted.index.isin(pathway)\n",
    "\n",
    "running_sum_storage = []\n",
    "running_sum = 0\n",
    "running_sum_storage.append(running_sum)\n",
    "n_in_pathway = np.sum(barcode)\n",
    "n_not_in_pathway = len(barcode) - np.sum(barcode)\n",
    "for ele in barcode:\n",
    "    if ele: # gene is in gene set\n",
    "        running_sum += 1/n_in_pathway\n",
    "    else: # gene is not in gene set\n",
    "        running_sum -= 1/n_not_in_pathway\n",
    "    running_sum_storage.append(running_sum)\n",
    "running_sum_storage = np.array(running_sum_storage)\n",
    "\n",
    "es_true = running_sum_storage[np.argmax(np.abs(running_sum_storage))]\n",
    "\n",
    "####### Permuting genes\n",
    "\n",
    "np.random.seed(1701)\n",
    "\n",
    "n_permutations = 10_000\n",
    "es_null_storage_array = []\n",
    "for permute_n in range(n_permutations):\n",
    "\n",
    "    barcode = tribble_df_sorted.index.isin(pathway)\n",
    "    new_barcode = np.random.permutation(barcode)\n",
    "\n",
    "    running_sum_storage = []\n",
    "    running_sum = 0\n",
    "    running_sum_storage.append(running_sum)\n",
    "    n_in_pathway = np.sum(barcode)\n",
    "    n_not_in_pathway = len(barcode) - np.sum(barcode)\n",
    "    for ele in new_barcode:\n",
    "        if ele: # gene is in gene set\n",
    "            running_sum += 1/n_in_pathway\n",
    "        else: # gene is not in gene set\n",
    "            running_sum -= 1/n_not_in_pathway\n",
    "        running_sum_storage.append(running_sum)\n",
    "    running_sum_storage = np.array(running_sum_storage)\n",
    "    # get value of max deviation from zero\n",
    "    es_null_storage_array.append(running_sum_storage[np.argmax(np.abs(running_sum_storage))])\n",
    "    \n",
    "es_null_storage_array = np.array(es_null_storage_array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(8,5))\n",
    "sns.distplot(es_null_storage_array, bins=np.arange(-1,1.01,0.1))\n",
    "\n",
    "ax = plt.gca()\n",
    "ax.axvline(es_true, color='r', label='true ES')\n",
    "ax.legend(fontsize=14)\n",
    "ax.set_xlabel('ES (null)', fontsize=16)\n",
    "ax.set_ylabel('density', fontsize=16)\n",
    "ax.tick_params(labelsize=14)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_es = es_null_storage_array[es_null_storage_array > 0] # get array of positive enrichment scores\n",
    "print('Empirical p-value: ')\n",
    "print(np.sum(pos_es >= es_true)/len(pos_es))\n",
    "\n",
    "nes = es_true/np.mean(np.abs(pos_es))\n",
    "print('NES: ')\n",
    "print(nes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that our null distribution of ES values is tighter than before, leaving our true ES even further outside the realm of expectation. Not a single one of our null ES values falls above our actual ES value! Our NES is also much higher. As with many other procedures in genomics, the outcome of our GSEA is highly dependent on the exact parameterization of our procedure.\n",
    "\n",
    "One other thing you should notice: we don't need any sample-specific data when performing GSEA with gene permutation! We only need the ranked list of fold-changes between the conditions for each gene."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4.2 GSEA with WebGestalt\n",
    "\n",
    "As with over-representation analysis, you are probably using GSEA on a large body of gene ontologies and pathways to determine which pathways are enriched. This is a slog to do manually, but our friend [WebGestalt](http://www.webgestalt.org/) makes it quite easy!\n",
    "\n",
    "WebGestalt is (currently) only capable of performing GSEA with gene permutation (also known as GSEA pre-ranked). As input, it takes tab-delimited file with genes in one column and fold-changes in the other (file extension `.rnk`). You can find a brief tutorial on how to use WebGestalt for GSEA here **(Note 2/18/22: I have not yet recorded this lecture! This will be the same lecture I give on 2/18/22 in class)**.\n",
    "\n",
    "Let's perform GSEA when comparing esophageal adenocarcinoma to normal esophagus. The cells below will import the expression data, as before, then create a file just showing the fold-change and gene name, which you can use as input in WebGestalt. When you generate a report, you'll notice some familiar-looking plots!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "de_df = pd.read_csv('E-MTAB-4054-query-results.tsv', sep='\\t', header=3)\n",
    "de_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "de_df_slice = de_df[[\"Gene Name\", \"'esophageal adenocarcinoma' vs 'normal' .foldChange\"]].dropna()\n",
    "de_df_slice.to_csv('eac_vs_normal_fc_0.rnk', index=False, header=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4.3 A few quick notes on GSEA\n",
    "\n",
    "### 4.3.1 \"Leading edge\" subset\n",
    "\n",
    "Sometimes in GSEA you will hear a set of genes referred to as the \"leading edge\" subset. If the overall enrichment score is positive, this is the set of genes that appear before the peak of our running enrichment score plot. If the enrichment score is negative, this is the set of genes that appear after the trough of the enrichment score plot. The leading edge subset, in theory, represents the set of genes that \"drive\" enrichment. The cells below identify the leading edge subset of the tribble comparison."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(1701)\n",
    "\n",
    "\n",
    "control_vals = tribble_df[sample_names[labels == 'control']].values\n",
    "treatment_vals = tribble_df[sample_names[labels == 'tachyon']].values\n",
    "\n",
    "# calculate log2 fold change (log2(mean(tachyon)/mean(control)))\n",
    "tribble_df['log2FC'] = np.log2(np.mean(treatment_vals, axis=1)/np.mean(control_vals, axis=1))\n",
    "\n",
    "tribble_df_sorted = tribble_df.sort_values('log2FC', ascending=False)\n",
    "\n",
    "barcode = tribble_df_sorted.index.isin(pathway)\n",
    "\n",
    "running_sum_storage = []\n",
    "running_sum = 0\n",
    "running_sum_storage.append(running_sum)\n",
    "n_in_pathway = np.sum(barcode)\n",
    "n_not_in_pathway = len(barcode) - np.sum(barcode)\n",
    "for ele in barcode:\n",
    "    if ele: # gene is in gene set\n",
    "        running_sum += 1/n_in_pathway\n",
    "    else: # gene is not in gene set\n",
    "        running_sum -= 1/n_not_in_pathway\n",
    "    running_sum_storage.append(running_sum)\n",
    "running_sum_storage = np.array(running_sum_storage)\n",
    "\n",
    "fig, axes = plt.subplots(nrows=3, ncols=1, sharex=True, figsize=(5,6), gridspec_kw={'height_ratios': [3,2, 1]})\n",
    "\n",
    "axes[0].plot(running_sum_storage, color='lime')\n",
    "axes[0].axvline(np.argmax(np.abs(running_sum_storage)), color='k', linestyle='dashed')\n",
    "axes[0].axhline(0.0, color='k', linewidth=0.5)\n",
    "axes[0].set_ylim([-np.max(np.abs(running_sum_storage))*1.1, np.max(np.abs(running_sum_storage))*1.1])\n",
    "axes[0].set_ylabel('enrichment score', fontsize=16)\n",
    "axes[1].plot(np.arange(300)+0.5, tribble_df_sorted['log2FC'].values)\n",
    "axes[1].axhline(0, color='k', linewidth=0.5)\n",
    "axes[1].set_ylabel(r'log$_2$(FC)', fontsize=16)\n",
    "axes[2].imshow(tribble_df_sorted.index.isin(pathway).reshape(-1,1).T, aspect='auto', cmap='binary')\n",
    "axes[2].set_yticks([])\n",
    "axes[2].set_xlabel('gene index', fontsize=16)\n",
    "axes[2].set_xlim([0,300])\n",
    "\n",
    "for ax in axes:\n",
    "    ax.tick_params(labelsize=14) # set the fontsize of the tick labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tribble_df_sorted[(np.arange(300) <= np.argmax(np.abs(running_sum_storage))) & \n",
    "                              (tribble_df_sorted.index.isin(pathway))]['log2FC']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.3.2 Ranking non-differentially-expressed genes\n",
    "\n",
    "Some GSEA users are uncomfortable ranking genes by fold-change, for a couple of reasons:\n",
    "* Genes with higher fold-changes do not necessarily affect the behavior of gene pathways to a greater degree than genes with lower fold-changes\n",
    "* Estimates of fold-change are highly subject to variation across experiments, and thus the rank of the list can be highly variable.\n",
    "* Estimates of fold-change are very sensitive to outliers (while phenotype permutation accounts for this, gene permutation does not).\n",
    "\n",
    "As a result, sometimes users will opt for a different approach. The two most common approaches are:\n",
    "* Rank by fold-change after excluding all non-differentially-expressed genes (only really works if number of up-regulated and down-regulated genes are relatively well-balanced)\n",
    "* Rank by log10(p-value) * sign of change (a gene with a p value of 0.05 and a fold change of -2.3 would have a score of -np.log10(0.05)(-1) = -1.3  and a gene with a p-value of 0.10 and a fold-change of 0.5 would have a score of -np.log10(0.1)(1) = 1.\n",
    "\n",
    "I won't comment on the validity of each of these approaches (each could be most appropriate depending on the circumstances) but you will see all of them employed fairly often!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Congratulations!\n",
    "## You have completed this lesson on GSEA!\n",
    "\n",
    "To practice the skills you've acquired thus far, complete the excercises below."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 3.1\n",
    "\n",
    "Perform GSEA to assess whether the set of genes concerning tribble reproduction (stored in `pathway` below) is up- or down-regulated after tachyon pulse exposure. Perform phenotype permutation and find a p-value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathway = np.array(['JMT3', 'XVI6', 'LJW1', 'HGE8', 'IRT4', 'DXN6', 'OVG3', 'OVG3',\n",
    "       'SMH2', 'DUK9', 'RNF4', 'CJY3', 'HBD9', 'ICP4', 'TAN8', 'RDC4',\n",
    "       'PDX4', 'JMT3', 'QVM1', 'XBE7', 'LFM7', 'RCX2', 'ELG9', 'JYP9',\n",
    "       'OIA6', 'ANG5', 'SNX2', 'XTF6', 'SOO4', 'NLW8', 'SGI6', 'SMH2',\n",
    "       'RSI4', 'CJY3', 'EDR2', 'CXM8', 'RST6', 'ZWG6', 'QCX9', 'OIA6',\n",
    "       'DDT5', 'ZAG1', 'HRP2', 'YDJ8', 'AZT9', 'MJS2', 'JPB9', 'LFA1',\n",
    "       'UDS8', 'TPD7'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 3.2\n",
    "\n",
    "The genes stores in `pathway` below represent the genes related to the Klingon defense pathway in tribbles. In the tachyon vs. control comparison, what genes in this pathway form the leading-edge subset?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathway = np.array(['GPS8', 'KRV5', 'CCO3', 'RST6', 'PKP1', 'QSV2', 'ERR1', \n",
    "                    'XPQ5', 'XVI6', 'UWD1', 'AKV4', 'QLG5', 'PNL2', 'XDP8', \n",
    "                    'XPQ5'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 3.3\n",
    "\n",
    "What pathways are the three most up-regulated and three most down-regulated (read: highest and lowest NES) in the lung during COVID-19? Rank your genes by fold-change (dropping any genes with NaN fold change) and use `Pathways > KEGG` as your functional database in your WebGestalt input. Upload the `.html` file of your report alongside your Jupyter Notebook (hit `Result Download`, and you will find the associated `.html` in the `.zip` file that downloads)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "de_df = pd.read_csv('E-ENAD-46-query-results.tsv', sep='\\t', header=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "de_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Appendix:\n",
    "\n",
    "See below for code used to create synthetic tribble data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_genes = 300\n",
    "n_replicates = 10\n",
    "np.random.seed(100)\n",
    "alpha = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'\n",
    "numeric = '123456789'\n",
    "df = pd.DataFrame({'gene_name':[], 'log2FC':[], 'FDR':[]})\n",
    "for gene in range(n_genes):\n",
    "    exp_val = np.random.normal(loc=0, scale=2)\n",
    "    df_slice = pd.DataFrame({'gene_name':[''.join(np.random.choice(list(alpha), 3)) + ''.join(np.random.choice(list(numeric), 1))],\n",
    "                 'log2FC':[exp_val], 'FDR':[10**(-(np.abs(exp_val)**2)*np.random.uniform(0.05,0.25))]})\n",
    "    df = df.append(df_slice)\n",
    "    \n",
    "df = df.reset_index(drop=True)\n",
    "\n",
    "baseline = 2**np.random.normal(loc=7, scale=3.5, size=n_genes)\n",
    "treatment = pd.DataFrame(2**np.random.normal(loc=df['log2FC'], scale=0.75, size=(n_replicates,n_genes)).T).mul(baseline, axis=0)\n",
    "control = pd.DataFrame(2**np.random.normal(loc=0, scale=0.75, size=(n_replicates,n_genes)).T).mul(baseline, axis=0)\n",
    "control.columns = ['control_' + str(x) for x in range(n_replicates)]\n",
    "treatment.columns = ['tachyon_' + str(x) for x in range(n_replicates)]\n",
    "labels = np.array(list(['control']*n_replicates) + list(['tachyon']*n_replicates))\n",
    "df[control.columns] = control\n",
    "df[treatment.columns] = treatment\n",
    "\n",
    "df = df[df.columns[(df.columns != 'log2FC') & (df.columns != 'FDR')]]\n",
    "df\n",
    "\n",
    "# Write to .csv (overwrite at your own risk!)\n",
    "#df.to_csv('tribble_expression_experiment_220217.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
